zerver/lib/upload.py

```diff
@@ -184,21 +183,22 @@ def upload_image_to_s3(
 
     key.set_contents_from_string(contents, headers=headers)  # type: ignore # https://github.com/python/typeshed/issues/1552
 
-def get_total_uploads_size_for_user(user: UserProfile) -> int:
-    uploads = Attachment.objects.filter(owner=user)
-    total_quota = uploads.aggregate(Sum('size'))['size__sum']
+def currently_used_upload_space(realm: Realm) -> int:
+    total_quota = Attachment.objects.filter(realm=realm).aggregate(Sum('size'))['size__sum']
 
-    # In case user has no uploads
+    # In case realm has no uploads
     if (total_quota is None):
         total_quota = 0
     return total_quota
 
-def within_upload_quota(user: UserProfile, uploaded_file_size: int) -> bool:
-    total_quota = get_total_uploads_size_for_user(user)
-    if (total_quota + uploaded_file_size > user.quota):
-        return False
-    else:
-        return True
+def check_upload_within_quota(realm: Realm, uploaded_file_size: int) -> None:
+    if realm.uploads_quota_gb is None:
+        return
+    used_space = currently_used_upload_space(realm)
+    # Convert realm.uploads_quota_gb to Bytes and check whether the upload
+    # will cross realm quota.
+    if (used_space + uploaded_file_size) > (realm.uploads_quota_gb << 30):
```

Comment:
> realm.uploads_quota_gb << 30

This is going to bug me, because a GB aka a gigabyte is not 2^30 bytes -- it is 10^9 bytes. The point of the international standard prefix "giga-" (and its abbreviation "G") is that it always means the same thing :smile: , and that meaning is "10^9".

There is a prefix for 2**30: "gibi", abbreviated "Gi", as in 1 GiB for 1 gibibyte, which equals 1024 MiB (or mebibytes), etc.

As a product matter, we should probably make the quota in gibibytes, because the difference is immaterial and it's a bit more generous. Basically a ["baker's gigabyte"](https://www.britannica.com/story/why-is-a-bakers-dozen-13). And though it pains me a bit, we should probably even call those gibibytes "GB" and even "gigabytes" in most user-facing text, because otherwise we have to explain what a gibibyte is. (We might explain somewhere what our "GB" actually are.)

But the code is the place for us to be super clear to ourselves about how things actually work. :grinning: So:
* Let's centralize this calculation in just one place: a method or property on `Realm` with a name like `uploads_quota_bytes`.
* In that function, have a comment explaining this reasoning: even though we say "GB", we use GiB to be more generous.

Reply:
Quick summary of some web research I did on that product question of whether it should actually be gibibytes:
* Dropbox's plans [are quoted](https://www.dropbox.com/plans) in what appear to be tebibytes (e.g. "1 TB (1024 GB)").
* They even [used to](https://github.com/owncloud/core/issues/13386#issuecomment-123645835) say "1 TB (1,000 GB)" -- apparently someone decided the baker's terabyte was a better idea. (I don't actually recall from my time at Dropbox what the quotas actually were, and whether the actual values changed or someone was just confused earlier.)
* Phones' storage sizes [seem](https://www.reddit.com/r/Android/comments/30myl5/samsung_galaxy_s6_actual_usable_storage/cpu0hg2/) to [be quoted](https://www.reddit.com/r/GooglePixel/comments/5m7err/have_a_32gb_pixel_describe_your_storage_space/dc1n3u3/) ([iPhones included](https://www.quora.com/How-much-storage-capacity-does-the-iPhone-6-Plus-64-GB-model-really-have)) in genuine GB. Then again, those links come mostly from people complaining about how the phone vendors are cheating them, because they don't understand it.
* [AWS S3](https://forums.aws.amazon.com/message.jspa?messageID=54595#54595), and [Google Cloud Storage](https://cloud.google.com/storage/pricing), both quote prices in "GB" that are actually GiB.